cmake/everest-generate.cmake

```diff
@@ -494,7 +499,82 @@ function (ev_add_py_module MODULE_NAME)
             add_subdirectory(${MODULE_PATH})
         endif()
     else()
-        message(WARNING "C++ module ${MODULE_NAME} does not exist at ${MODULE_PATH}")
+        message(WARNING "Python module ${MODULE_NAME} does not exist at ${MODULE_PATH}")
+        return()
+    endif()
+
+    # FIXME (aw): this will override EVEREST_MODULES, might not what we want
+    set_property(
+        GLOBAL
+        PROPERTY EVEREST_MODULES ${EVEREST_MODULES}
+    )
+endfunction()
+
+function (ev_add_rs_module MODULE_NAME)
+    set(EVEREST_MODULE_INSTALL_PREFIX "${CMAKE_INSTALL_LIBEXECDIR}/everest/modules")
+
+    set(MODULE_PATH "${CMAKE_CURRENT_SOURCE_DIR}/${MODULE_NAME}")
+    set(MODULE_BUILD_PATH "${CMAKE_CURRENT_BINARY_DIR}/${MODULE_NAME}")
+
+    if(IS_DIRECTORY ${MODULE_PATH})
+        if(NOT ${EVEREST_ENABLE_RS_SUPPORT})
+            message(STATUS "Excluding Rust module ${MODULE_NAME} because EVEREST_ENABLE_RS_SUPPORT=${EVEREST_ENABLE_RS_SUPPORT}")
+            return()
+        elseif("${MODULE_NAME}" IN_LIST EVEREST_EXCLUDE_MODULES)
+            message(STATUS "Excluding module ${MODULE_NAME}")
+            return()
+        elseif(EVEREST_INCLUDE_MODULES AND NOT ("${MODULE_NAME}" IN_LIST EVEREST_INCLUDE_MODULES))
+            message(STATUS "Excluding module ${MODULE_NAME}")
+            return()
+        else()
+            message(STATUS "Setting up Rust module ${MODULE_NAME}")
+
+            find_program(
+                    RSYNC
+                    rsync
+                    REQUIRED
+            )
+
+            # Copy source of module ${MODULE_NAME} to build dir
+            execute_process(
+                COMMAND ${RSYNC} -avq --delete "${MODULE_PATH}/" "${MODULE_BUILD_PATH}"
+            )
+
+            find_program(
+                SED
+                sed
+                REQUIRED
+            )
```

Comment:
As discussed, but only for the cmake build flow:
I would create a `build_rust_modules` target, which accesses a list (this list gets appended to via the ev_add_rust_module function) of rust modules.  This list then gets serialized into a generated cargo workspace file, which then should be build.
General problems here are how to track changes in rust source code.  One possibility would be in a rs module folder to add the files manually to a dependency list, which then gets checked.  But this doesn't really scale ...
On the other side, we could always run cargo build and let cargo determine if something new needs to be compiled.
But there are still a lot of other issues, I guess.

Reply:
@hikinggrass I do not think this is the right approach here - the rsync is rather brutal and makes it hard to track changes. I think @a-w50 idea has merits. I also think I found a solution that works - it relies on the ability to specify dependencies in the Workspace Cargo.toml and have members use that. The idea is outlined here: https://github.com/SirVer/rust_cmake_build_outline

1. The idea is that the sources must all be in a cargo workspace. And they do not reference everstrs directly, instead they use the `everstrs = { workspace = true }`. The source workspace [Cargo.toml](https://github.com/SirVer/rust_cmake_build_outline/blob/main/source/Cargo.toml) references them. 
2. For build, we create a new workspace in the build dir, we symlink all the modules and the everstrs source dir into this and create a [Cargo.toml](https://github.com/SirVer/rust_cmake_build_outline/blob/main/build/Cargo.toml) that references the everstrs that we care about. 
3. We then use https://github.com/corrosion-rs/corrosion to build this workspace with cMake, this should give us perfect reproducability and dependency tracking.

What do you think? @a-w50 @hikinggrass

Reply:
I did implement this now. Seems to be working fine.